query-algebrizer/src/clauses/fulltext.rs

```diff
@@ -152,64 +169,46 @@ impl ConjoiningClauses {
+            self.bind_column_to_var(schema, fulltext_values_alias.clone(), Column::Fulltext(FulltextColumn::Text), var.clone());
         }
 
-        if let VariableOrPlaceholder::Variable(ref var) = bindings[2] {
-            if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
-                // It's OK for this to collide: the two must unify.
-                unimplemented!();
-            }
+        if let VariableOrPlaceholder::Variable(ref var) = b_tx {
             self.constrain_var_to_type(var.clone(), ValueType::Ref);
+            if self.is_known_empty() {
+                return Ok(());
+            }
 
-            let tx_alias = QualifiedAlias(datoms_table_alias.clone(), Column::Fixed(DatomsColumn::Tx));
-            self.column_bindings.entry(var.clone()).or_insert(vec![]).push(tx_alias);
+            self.bind_column_to_var(schema, datoms_table_alias.clone(), DatomsColumn::Tx, var.clone());
         }
 
-        if let VariableOrPlaceholder::Variable(ref var) = bindings[3] {
-            if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
-                // It's OK for this to collide: the two must unify.
-                unimplemented!();
-            }
+        if let VariableOrPlaceholder::Variable(ref var) = b_score {
             self.constrain_var_to_type(var.clone(), ValueType::Double);
 
-            // Right now we don't support binding a column to a constant value.  Therefore, we
-            // introduce a computed table with exactly the constant value we require.  See the
-            // translate tests for some edge cases in this space.
-            // TODO: optimize this by binding the value, like:
-            // self.value_bindings.insert(var.clone(), TypedValue::Double(0.0.into()));
+            // We do not allow the score to be bound.
+            if self.value_bindings.contains_key(var) || self.input_variables.contains(var) {
+                bail!(ErrorKind::InvalidBinding(var.name(), BindingError::UnexpectedBinding));
+            }
+
+            // We bind the value ourselves. This takes care of substituting into existing uses.
             // TODO: produce real scores using SQLite's matchinfo.
-            let names: Vec<Variable> = vec![var.clone()];
-            let named_values = ComputedTable::NamedValues {
-                names: names.clone(),
-                values: vec![TypedValue::Double(0.0.into())],
-            };
-            let table = self.computed_tables.push_computed(named_values);
-            let alias = self.next_alias_for_table(table);
-
-            // Stitch the computed table into column_bindings, so we get cross-linking.
-            self.bind_column_to_var(schema, alias.clone(), VariableColumn::Variable(var.clone()), var.clone());
-
-            self.from.push(SourceAlias(table, alias.clone()));
+            self.bind_value(var, TypedValue::Double(0.0.into()));
```

Comment:
Ah, very nice.  I never found the expression for this idea while working on these patches.